[clang][WebAssembly] Handle casted function pointers with different number of arguments

turran opened the pull request:
This overcomes the limitation of WebAssemblyFixFunctionBitcasts LLVM module after the introduction of Opaque Pointers. The function pointers no longer have a prototype; therefore, the solution has to be done at clang. This solves many open issues, like:

* https://github.com/llvm/llvm-project/issues/64526
* https://github.com/emscripten-core/emscripten/issues/23952
* https://gitlab.gnome.org/GNOME/glib/-/issues/3596

turran committed:
[wasm] Support different signature function pointers

kleisauke committed:
Prefer use of Wasm signatures when building the thunk name

turran committed:
[wasm] Add a thunk cache

turran committed:
[wasm] Add function pointer thunk tests

juj commented:
Thanks for working on this.

If I understand correctly, this will silently paper over function arg count mismatches in signature? That might help some code patterns, though there is a potential issue of silent code size bloat, if the developer did not intend to write a mismatching signature in the first place(?) (and if they didn't see it, Clang will then silently generate these thunks, which were not desired?)

In our use case of #64526, I still wish there was a diagnostics feature that could be used like `-Wbad-wasm-function-cast` and `-Werror=bad-wasm-function-cast` where Clang would diagnose those casts from->to a bad function signature. So this PR does not quite address #64526 it seems.

efriedma-quic reviewed (changes requested):
The C++ standard has exactly one rule for the result of function pointer casts: "converting a prvalue of type 'pointer to T1' to the type 'pointer to T2' (where T1 and T2 are function types) and back to its original type yields the original pointer value".  This trivially violates that rule.  (C has a similar rule.)

So we can't do this.  At least, not by default.

turran commented:
Thanks @efriedma-quic 

> The C++ standard has exactly one rule for the result of function pointer casts: "converting a prvalue of type 'pointer to T1' to the type 'pointer to T2' (where T1 and T2 are function types) and back to its original type yields the original pointer value". This trivially violates that rule. (C has a similar rule.)
> 
> So we can't do this. At least, not by default.

I see. I wasn't aware of this rule. What would be the preferred way to handle this, not by default, with a specific flag, I guess?

juj commented:
What I would like is a new flag exactly like the current `-Wcast-function-type`, but one that was just narrower in its emitted warnings.

I.e. currently `-Wcast-function-type` emits warnings for all sorts of casts that are safe with Wasm ABI (like casting between `unsigned int` to `int` in a function arg).. but I'd like to not get warnings for those. And only get warnings for a narrower set of casts that will cause a trap at Wasm runtime. That would help audit through the millions of lines of code that e.g. Unity3D engine compiles through. (`-Wcast-function-type` is too much, and `-Wbad-function-cast` is not enough, as per illustrated in #64526)

Like was pointed out, of course detecting all scenarios that result to Wasm VM trapping at runtime is not possible (e.g. casting `void(int,int) -> void* -> void(int)` and then attempting to call), but that is besides the point. Because we couldn't detect all such scenarios doesn't mean it wouldn't be extremely useful to diagnose all the scenarios that can reasonably be compile-time diagnosed.

juj commented:
In our case, I am interested in getting warnings emitted. So the usual `-Wfoo`, `-Werror=foo` mechanism would be useful.

In this PR, the idea of transparently supporting the cast via a thunk was brought up, that might be something to consider under a `-fwasm-...` flag kind of behavior, though would be good to be alternative to the emit warnings/emit errors mode.

turran commented:
> Maybe a flag. But I'd prefer to explore the design space first. What patterns do we care about? Can we detect all the relevant cases at compile-time? Can we generalize this by performing some checks at runtime?

I added a couple of tests that highly represent the main set of use cases that this PR solves.  Using a cast for a function pointer struct field declaration, and when using a cast for call arguments.

About other potential relevant cases, I guess there might be more, but for my own scenario with the PR, it is enough.
Let me know if the design is reasonable for these cases

turran commented:
> What I would like is a new flag exactly like the current `-Wcast-function-type`, but one that was just narrower in its emitted warnings.
> 
> I.e. currently `-Wcast-function-type` emits warnings for all sorts of casts that are safe with Wasm ABI (like casting between `unsigned int` to `int` in a function arg).. but I'd like to not get warnings for those. And only get warnings for a narrower set of casts that will cause a trap at Wasm runtime. That would help audit through the millions of lines of code that e.g. Unity3D engine compiles through. (`-Wcast-function-type` is too much, and `-Wbad-function-cast` is not enough, as per illustrated in #64526)
> 
> Like was pointed out, of course detecting all scenarios that result to Wasm VM trapping at runtime is not possible (e.g. casting `void(int,int) -> void* -> void(int)` and then attempting to call), but that is besides the point. Because we couldn't detect all such scenarios doesn't mean it wouldn't be extremely useful to diagnose all the scenarios that can reasonably be compile-time diagnosed.

Okay, I've re-read the issue you mentioned. Indeed, it makes sense to provide a diagnosis of such use cases. Maybe as part of another PR? This was more in the direction of a different number of arguments instead of a different type of arguments.

juj commented:
> This was more in the direction of a different number of arguments instead of a different type of arguments.

I see these would be good to be handled as the same thing: i.e. a signature mismatch is a mismatch, be it types or counts of args? That is how the existing `-Wcast-function-type` and `-Wbad-function-cast` also work like.

efriedma-quic reviewed (commented):
Really, I think the biggest problem with this approach is that it special-cases compile-time casts, and treats them differently from runtime casts.  That might be just barely sufficient for specific cases from glib, but it's not comprehensive, and it's very hard for users to understand why specific cases are/are not handled.

The semantics would be much simpler if the code for performing indirect calls was changed.  Instead of trying to fix function pointers when they're cast, fix them when something tries to call them.  Some kind of table lookup, or something with ref.test.  Not sure exactly what this looks like; I'm not that familiar with wasm.

turran commented:
> Really, I think the biggest problem with this approach is that it special-cases compile-time casts, and treats them differently from runtime casts. That might be just barely sufficient for specific cases from glib, but it's not comprehensive, and it's very hard for users to understand why specific cases are/are not handled.
> 

I understand your point, but it is difficult to find a compromise in this situation. In this PR, specific -handable cases- are transformed at compile time to avoid a SEGV, which looks reasonable. Developers have pushed this function pointer casting (un)defined behavior because it has always worked in other targets, but not in WebAssembly. Maybe other cases are missing and need to be added?

> The semantics would be much simpler if the code for performing indirect calls was changed. Instead of trying to fix function pointers when they're cast, fix them when something tries to call them. Some kind of table lookup, or something with ref.test. Not sure exactly what this looks like; I'm not that familiar with wasm.

Doing a runtime fix like using `ref.test` would mean that every function pointer called must be first checked, for every combination of fewer arguments (calling a function pointer of three args would require a runtime check for two args, one arg and no args), transferring the complexity to runtime and adding this penalty when you know beforehand at compilation time how you are casting the function pointer. This is a considerable overhead.

A LUT approach would mean that for every `ptr` called as a function, it is needed to traverse back such a pointer to know what value it has been set to, confirm that it is callable, and, if not, change the value set to a thunk. I'm not sure if it is actually doable at the LLVM IR level. I'm not an expert on that.

Initially, the code at https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/WebAssembly/WebAssemblyFixFunctionBitcasts.cpp was doing a similar approach as the one found on this PR, but after the Opaque Pointer change, the function pointer information found on LLVM's IR is only a generic pointer and such an approach doesn't scale anymore.

turran committed:
[wasm] Add -fwasm-fix-function-bitcasts to enable the thunk generation

turran commented:
> Or I guess we can go with the current design, behind a flag that indicates you understand what you're doing, with the understanding that it isn't standard-compliant, and only handles specifically the patterns that show up in some subset of glib. Which is not at all pretty, but I guess it doesn't break anyone else.

I added a new flag, `-fwasm-fix-function-bitcasts,` to enable the proposed behavior. I've been struggling with the implementation, and I'm not sure it is well done. The WASM-specific flags differ from other targets, so I preferred to be somewhat compatible with the others. Let me know if the implementation is correct.

turran commented:
> Re-reading back the original description of the PR, I see there's
> 
> > This solves many open issues, like:
> > ```
> > * https://github.com/llvm/llvm-project/issues/64526
> > ```
> 
> though this PR doesn't quite solve this, since it (silently?) will try to fix up the call sites with mismatching pointers. In issue 64526, I am looking after a way to get a warning or an error at compile time for call sites that perform casts that are unsafe specifically for Wasm.
> 
> In other words, in production I am not sure if I would like to use `-fwasm-fix-function-bitcasts` to "hide the problem under the rug", but I would like to have a more focused warning to guide analysing casts that are problematic for WebAssembly, with the intent of fixing the code of such call sites by hand.
> 
> Of course it is a fantastic idea to have a build option that can work on these automatically. I am sure that some use cases will find it very useful.

Indeed. I'm aware of your particular case and have some clue on where/how to do it. It will touch other parts as it needs to modify the SemaCast.cpp and improve the WebAssemblyABIInfo. If you prefer, I can remove your issue from the description. I'll need some time to draft something, as compiling/linking/testing, at least for me, is very tedious.

efriedma-quic commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4932,6 +4932,27 @@ void CodeGenFunction::EmitCallArg(CallArgList &args, const Expr *E,
+  // For WebAssembly target we need to create thunk functions
+  // to properly handle function pointers args with a different signature.
+  // Due to opaque pointers, this can not be handled in LLVM
+  // (WebAssemblyFixFunctionBitcast) anymore
```
Talking about opaque pointers isn't really useful for a code comment, because it's not about the current state of the code.

efriedma-quic reviewed (commented):
I don't think I have any further comments on the code... but I'm not familiar enough with wasm to judge whether this is actually worth adding.

@dschuff @aheejin Thoughts?

dschuff commented:
> Hello @dschuff. Thanks for your comments.
> 
> > Or another random idea (for the "semi-transparently call with signature mismatches" problem, not the "fix the bugs in my code" problem): What if we represented function pointers (in the source) as being in a different address space, using the existing `address_space(N)` clang/LLVM attributes? They could be freely casted, written into memory and have the same integer values as they otherwise would, but when you call through them, we could either use a generic catch-all thunk like Binaryen's `fpcast-emu` pass does today, or perhaps something more fancy, like a LUT check and a more-specific conversion thunk. That could be feasible if e.g. the `FunctionPointer` C type and `FUNCTION_POINTER` macro could be suitably redefined. Then we maybe wouldn't need any changes to the code at callsites.
> 
> I'm not aware of what `address_space` does, but I understand it is something that would require a change in the original source code, which I think the approach should be transparent to the programmer.
> 
> About the LUT, isn't it similar to the approach commented on [#153168 (review)](https://github.com/llvm/llvm-project/pull/153168#pullrequestreview-3122024433) and replied on [#153168 (comment)](https://github.com/llvm/llvm-project/pull/153168#issuecomment-3212203688) ?

Let me refine what I meant here a little bit. The key problem I have with your proposed approach is that it redefines the semantics of the language in a subtle and hard-to-understand way (i.e. compile-time vs runtime casts, as mentioned [above](https://github.com/llvm/llvm-project/pull/153168#pullrequestreview-3122024433)). I think that this isn't going to be very "transparent to the programmer" if you have a flag that just changes this behavior across the program, it could have effects outside of the very-specially-designated intended uses.

I agree with the statement above that changing the indirect call behavior is better than changing the casting behavior because it's much easier to understand and more predictable. It's also why I'm wondering if we can localize the effect to specific callsites which can mitigate both any increased cost for the calls, and any potentially-surprising behavior changes. The example you've given have centralized macros such as `FUNCTION_POINTER` and typedefs for the various types involved. If we can, for example, define the macro such that the calls can call a compiler builtin to make the dispatch call, then we could localize any extra runtime costs. Or, if we could use the typedef to designate the pointer types in a special way, we could do something similar (this is what I was speculating about with the addrspace thing). The point is that the special calls or casts could be explicit rather than implicit. It's true that this isn't completely transparent in that it would require no source changes at all, but I think it's pretty reasonable, and I would hope that the GLib code is organized in such a way that only the special signal dispatch code needs this behavior, and it might therefore be acceptable to take some macro changes.

turran commented:
> 
> Let me refine what I meant here a little bit. The key problem I have with your proposed approach is that it redefines the semantics of the language in a subtle and hard-to-understand way (i.e. compile-time vs runtime casts, as mentioned [above](https://github.com/llvm/llvm-project/pull/153168#pullrequestreview-3122024433)). I think that this isn't going to be very "transparent to the programmer" if you have a flag that just changes this behavior across the program, it could have effects outside of the very-specially-designated intended uses.
>

I agree, but as I already commented, finding a middle ground, even if desirable, is sometimes impossible.
 
> I agree with the statement above that changing the indirect call behavior is better than changing the casting behavior because it's much easier to understand and more predictable. It's also why I'm wondering if we can localize the effect to specific callsites which can mitigate both any increased cost for the calls, and any potentially-surprising behavior changes.

The cost is unavoidable if the developer wants the webassembly ABI to work like other ABIs, where you simply pass fewer arguments than the ones expected, and don't change their own code. The proposed approach mimics such behavior; indeed, some collateral potential problems might occur, but that's the purpose of having an optional flag and knowing the implications.

> The example you've given have centralized macros such as `FUNCTION_POINTER` and typedefs for the various types involved. If we can, for example, define the macro such that the calls can call a compiler builtin to make the dispatch call, then we could localize any extra runtime costs. Or, if we could use the typedef to designate the pointer types in a special way, we could do something similar (this is what I was speculating about with the addrspace thing).

Yes, my examples have such a macro to be explicit about usage. But in the end, you can have a common macro like my example or a manual cast of the target function to the desired type in many ways and forms, which makes your approach developer-related, as they will have to know how to cast to make their cast to work. Maybe we can make the flag to be more explicit about the potential consequences?

>The point is that the special calls or casts could be explicit rather than implicit. It's true that this isn't completely transparent in that it would require no source changes at all, but I think it's pretty reasonable, and I would hope that the GLib code is organized in such a way that only the special signal dispatch code needs this behavior, and it might therefore be acceptable to take some macro changes.

We already tried some alternatives to avoid this problem (intermediate functions, different casts, adding missing parameters, etc), but those weren't accepted. In fact, there is a [hard requirement](https://gitlab.gnome.org/GNOME/glib/-/blob/main/docs/toolchain-requirements.md?ref_type=heads#calling-functions-through-differently-typed-function-pointers) in their documentation. This is mandatory at the compiler level. Nevertheless, there is a misunderstanding in the documentation about the `EMULATE_FUNCTION_POINTER_CASTS` as it has nothing to do with the number of arguments, only with the _type_ of arguments.

The signals are not the problem; is the whole GOBject inheritance model that is based on this ABI assumption.  It is difficult for me to find a middle ground here ...

dschuff commented:
> The cost is unavoidable if the developer wants the webassembly ABI to work like other ABIs, where you simply pass fewer arguments than the ones expected, and don't change their own code. The proposed approach mimics such behavior; indeed, some collateral potential problems might occur, but that's the purpose of having an optional flag and knowing the implications.

This is part of what I'm trying to understand about the use case. Namely, exactly who has to know the implications? (equivalently, who would need to enable this flag)? is it just the GLib core developers, i.e. the flag just needs to be enable to build GLib code? Or is it everyone who wants to link against GLib?

> The signals are not the problem; is the whole GOBject inheritance model that is based on this ABI assumption. It is difficult for me to find a middle ground here ...

This phrasing suggests that everyone who wants to use GLib has calls in their code (or at least their object files) that need to use mismatched signatures. The fact that IIUC you were able to make things work with hacks that were localized to GLib means that your changes just leaked into public headers?
 

> Nevertheless, there is a misunderstanding in the documentation about the EMULATE_FUNCTION_POINTER_CASTS as it has nothing to do with the number of arguments, only with the type of arguments.

EMULATE_FUNCTION_POINTER_CASTS (implemented by Binaryen's [fpcast-emu](https://github.com/WebAssembly/binaryen/blob/main/src/passes/FuncCastEmulation.cpp) transformation pass) makes all indirect calls use the same number of arguments, and creates thunks that pass those calls through with the right number of arguments. So it should fix both the number and type of arguments.

That was actually one question I had for you about the patch you wrote here vs EMULATE_FUNCTION_POINTER_CASTS, namely whether EMULATE_FUNCTION_POINTER_CASTS actually worked and was sufficient for GLib (it was unclear from the discussion; also a side question was if so, how bad was the performance effect?). It should work for both number and type of arguments. IIRC it only works for indirect calls, but we could probably fix that on the Binaryen side.

edit: for that matter, was the old FixFunctionBitcasts pass from before opaque pointers sufficient? I think not, right?

dschuff commented:
(BTW I'm going to be out for the next week, and then at the Wasm CG meeting the following week, so I probably won't get back to this until after that. I will get back to it though, despite my pushback here, I still want to come up with something here that works.).

dschuff commented:
Indeed, this is why I'm hoping we can make something explicit and targeted like a compiler builtin, rather than having to blindly transform the whole program.

turran committed:
[wasm] Add a more general approach at bitcast

turran commented:
Hello @dschuff
I'd like to continue discussing this PR and provide you some updates on the current status after rebasing with the current `main`.

Continuing with the `EMULATE_FUNCTION_POINTER_CASTS` approach, this solution is still not enough to have
a complete solution for the undefined behavior in function pointer type-mismatch cases. For complex programs, it simply does not work. Also, there are other correlated issues like:

* https://github.com/emscripten-core/emscripten/issues/15081
* https://github.com/emscripten-core/emscripten/issues/22285

Despite the performance penalty, I already commented on my previous post.

Getting back to the main concern of adding this to LLVM, I'd like to address each of the different points raised

### On C standard compliance

It is correct that the current thunk substitution breaks the round-trip identity guarantee, but there are several existing mechanisms on LLVM to deliberately deviate from the standard compliance, always through an explicit opt-in flag. For example, `-fwrapv`, `-ffast-math`, `-fno-strict-aliasing`, etc. I'd like to point out that a no-longer-working mechanism was already accepted based on the IR type information (`WebAssemblyFixFunctionBitcasts.cpp`). This PR restores that behavior at the Clang level, where type information is still available. Also, the behavior is gated behind `-fwasm-fix-function-bitcasts`, so no code is affected unless the developer explicitly opts in, fully aware of the implications. Keeping the overhead proportional to the actual problem.

### On compile-time vs. runtime casts

This is correct, and it is by design. The goal is not to handle every possible case of undefined behavior; it is to handle the specific, statically visible patterns.

Since the original `WebAssemblyFixFunctionBitcasts` pass, the scope was always limited to cases that are statically detectable. The opaque pointer change made those cases invisible at the IR level; this PR moves the detection back to the AST level in Clang, where the type information is preserved. 

### On coverage and the three code paths

I have extended the tests to include the `void *` cast too, making the implementation more generic now. I have created four test cases that might need to be extended for other scenarios, but so far, these cover the patterns that appear in real-world libraries like GLib/GObject.

As I mentioned before, I understand there might be other theoretical solutions, but no such solution exists today that covers the struct field case, for example, or limits the costs of doing an indirect call through a thunk. The choice is between an imperfect compile-time solution that covers the patterns found in GLib/GObject and similar C callback-heavy libraries, or no solution at all. The flag makes the trade-off explicit and opt-in.

Let me know your thoughts

turran committed:
[wasm] Add runtime wrapper for function pointer casts

This complements the existing compile-time thunk generation for statically
traceable function references.

turran committed:
[wasm] Match the static bitcast logic to the runtime one

turran committed:
wasm: Improve the runtime casts

Use a direct wrapper for immediate casts and a set of wrappers
for saved casted function pointers. Include a cache mechamism
and a documentation for the whole approach.

kleisauke commented on clang/lib/CodeGen/Targets/WebAssembly.cpp:
```diff
@@ -170,8 +324,504 @@ RValue WebAssemblyABIInfo::EmitVAArg(CodeGenFunction &CGF, Address VAListAddr,
+// See getInvokeSig() in WebAssemblyAsmPrinter for related logic.
+char WebAssemblyTargetCodeGenInfo::getTypeSig(const QualType &Ty,
+                                              const ASTContext &Ctx) const {
+  if (Ty->isAnyPointerType()) {
```
It looks like `isAnyPointerType()` doesn't catch C++ pointer types, so perhaps this should be:
```suggestion
  if (Ty->isCompoundType()) {
```

While this isn't reproducible with the current changes in this PR, it came up during the effort to sync `getOrCreateWasmFunctionPointerThunk()` with the `WebAssemblyFixFunctionBitcasts` pass.

<details>
  <summary>Details</summary>

Tested with these changesets:
https://github.com/fluendo/llvm-project/compare/218e499fe206f02894954bbd5aa5ddb3e5e78a48...kleisauke:wasm-function-pointers-rebase6
https://github.com/kleisauke/wasm-vips/compare/1740576a2cdcdd3a31ed6c4a370cd4e357554819...llvm-pr-153168

```console
FAILED: [code=262] gobject/libgobject-2.0.a.p/gsourceclosure.c.o 
emcc -Igobject/libgobject-2.0.a.p -Igobject -I../gobject -I. -I.. -Iglib -I../glib -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=gnu99 -O3 -D_GNU_SOURCE -fno-strict-aliasing -DG_DISABLE_CAST_CHECKS -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -Wfloat-conversion -Wimplicit-fallthrough -Wmisleading-indentation -Wmissing-field-initializers -Wnonnull -Wnull-dereference -Wunused -Wno-unused-parameter -Wno-cast-function-type -Wno-pedantic -Wno-format-zero-length -Wno-variadic-macros -Werror=format=2 -Werror=init-self -Werror=missing-include-dirs -Werror=pointer-arith -Werror=unused-result -Wstrict-prototypes -Wno-bad-function-cast -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=pointer-sign -Wno-string-plus-int -Wno-typedef-redefinition -Os -pthread -fwasm-exceptions -fvisibility=hidden -fwasm-fix-function-bitcasts -msimd128 -DWASM_SIMD_COMPAT_SLOW -fPIC '-DG_LOG_DOMAIN="GLib-GObject"' -DGOBJECT_COMPILATION -Wsign-conversion -Wshorten-64-to-32 -MD -MQ gobject/libgobject-2.0.a.p/gsourceclosure.c.o -MF gobject/libgobject-2.0.a.p/gsourceclosure.c.o.d -o gobject/libgobject-2.0.a.p/gsourceclosure.c.o -c ../gobject/gsourceclosure.c
../gobject/gsourceclosure.c:82:1: warning: unused function 'io_watch_closure_callback' [-Wunused-function]
   82 | io_watch_closure_callback (GIOChannel   *channel,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
Unhandled QualType
UNREACHABLE executed at /home/kleisauke/emsdk/llvm/git/src/clang/lib/CodeGen/Targets/WebAssembly.cpp:707!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang -target wasm32-unknown-emscripten -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -wasm-use-legacy-eh -mllvm -disable-lsr --sysroot=/home/kleisauke/emsdk/emscripten/main/cache/sysroot -D__EMSCRIPTEN_SHARED_MEMORY__=1 -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -Igobject/libgobject-2.0.a.p -Igobject -I../gobject -I. -I.. -Iglib -I../glib -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=gnu99 -O3 -D_GNU_SOURCE -fno-strict-aliasing -DG_DISABLE_CAST_CHECKS -DG_DISABLE_ASSERT -DG_DISABLE_CHECKS -Wfloat-conversion -Wimplicit-fallthrough -Wmisleading-indentation -Wmissing-field-initializers -Wnonnull -Wnull-dereference -Wunused -Wno-unused-parameter -Wno-cast-function-type -Wno-pedantic -Wno-format-zero-length -Wno-variadic-macros -Werror=format=2 -Werror=init-self -Werror=missing-include-dirs -Werror=pointer-arith -Werror=unused-result -Wstrict-prototypes -Wno-bad-function-cast -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=pointer-sign -Wno-string-plus-int -Wno-typedef-redefinition -Os -pthread -fwasm-exceptions -fvisibility=hidden -fwasm-fix-function-bitcasts -msimd128 -DWASM_SIMD_COMPAT_SLOW -fPIC -DG_LOG_DOMAIN=\"GLib-GObject\" -DGOBJECT_COMPILATION -Wsign-conversion -Wshorten-64-to-32 -MD -MQ gobject/libgobject-2.0.a.p/gsourceclosure.c.o -MF gobject/libgobject-2.0.a.p/gsourceclosure.c.o.d -ogobject/libgobject-2.0.a.p/gsourceclosure.c.o -c ../gobject/gsourceclosure.c
1.	<eof> parser at end of file
2.	Per-file LLVM IR generation
3.	../gobject/gsourceclosure.c:32:1 <Spelling=<scratch space>:46:1>: Generating code for declaration 'g_io_channel_get_type_once'
 #0 0x00000000024ea2e1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x24ea2e1)
 #1 0x00000000024e7ba7 llvm::sys::CleanupOnSignal(unsigned long) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x24e7ba7)
 #2 0x0000000002418144 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007efe29d0a0b0 __restore_rt (/lib64/libc.so.6+0x1a0b0)
 #4 0x00007efe29d64dcc __pthread_kill_implementation (/lib64/libc.so.6+0x74dcc)
 #5 0x00007efe29d09f8e gsignal (/lib64/libc.so.6+0x19f8e)
 #6 0x00007efe29cf17b3 abort (/lib64/libc.so.6+0x17b3)
 #7 0x0000000002423dbc llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2423dbc)
 #8 0x0000000002c8fc5d (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2c8fc5d)
 #9 0x0000000002c92271 WebAssemblyTargetCodeGenInfo::getThunkName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>, clang::FunctionProtoType const*, clang::ASTContext const&) const (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2c92271)
#10 0x0000000002c927d5 WebAssemblyTargetCodeGenInfo::getOrCreateWasmFunctionPointerThunk(clang::CodeGen::CodeGenModule&, llvm::Value*, clang::QualType, clang::QualType) const (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2c927d5)
#11 0x000000000289fe2a (anonymous namespace)::ScalarExprEmitter::VisitCastExpr(clang::CastExpr*) CGExprScalar.cpp:0:0
#12 0x000000000289c59f clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x289c59f)
#13 0x0000000002d815a4 clang::CodeGen::CodeGenFunction::EmitScalarInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2d815a4)
#14 0x0000000002d82f70 clang::CodeGen::CodeGenFunction::EmitExprAsInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2d82f70)
#15 0x0000000002d8c3c1 clang::CodeGen::CodeGenFunction::EmitAutoVarInit(clang::CodeGen::CodeGenFunction::AutoVarEmission const&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2d8c3c1)
#16 0x0000000002d91079 clang::CodeGen::CodeGenFunction::EmitVarDecl(clang::VarDecl const&) (.part.0) CGDecl.cpp:0:0
#17 0x0000000002d9180d clang::CodeGen::CodeGenFunction::EmitDecl(clang::Decl const&, bool) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2d9180d)
#18 0x00000000029b24d4 clang::CodeGen::CodeGenFunction::EmitDeclStmt(clang::DeclStmt const&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x29b24d4)
#19 0x00000000029c73c5 clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x29c73c5)
#20 0x00000000029be961 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x29be961)
#21 0x00000000029c6ec3 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x29c6ec3)
#22 0x0000000002a2f09c clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2a2f09c)
#23 0x0000000002a46054 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2a46054)
#24 0x0000000002aa8270 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2aa8270)
#25 0x0000000002aa3b94 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2aa3b94)
#26 0x0000000002aaf19e clang::CodeGen::CodeGenModule::EmitDeferred() (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2aaf19e)
#27 0x0000000002ab134e clang::CodeGen::CodeGenModule::Release() (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2ab134e)
#28 0x0000000002e46dac (anonymous namespace)::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) ModuleBuilder.cpp:0:0
#29 0x0000000002e43ff5 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2e43ff5)
#30 0x0000000004cb5df4 clang::ParseAST(clang::Sema&, bool, bool) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x4cb5df4)
#31 0x00000000031d94de clang::FrontendAction::Execute() (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x31d94de)
#32 0x000000000314f03b clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x314f03b)
#33 0x00000000032e8c92 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x32e8c92)
#34 0x0000000000d3d8af cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0xd3d8af)
#35 0x0000000000d35998 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>) driver.cpp:0:0
#36 0x0000000000d35b7a int llvm::function_ref<int (llvm::SmallVectorImpl<char const*>&)>::callback_fn<clang_main(int, char**, llvm::ToolContext const&)::'lambda'(llvm::SmallVectorImpl<char const*>&)>(long, llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#37 0x0000000002eb8d39 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#38 0x000000000241854b llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x241854b)
#39 0x0000000002eb95cd clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#40 0x0000000002e748ef clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2e748ef)
#41 0x0000000002e758ef clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2e758ef)
#42 0x0000000002e83815 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0x2e83815)
#43 0x0000000000d3ad93 clang_main(int, char**, llvm::ToolContext const&) (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0xd3ad93)
#44 0x0000000000c10654 main (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0xc10654)
#45 0x00007efe29cf3681 __libc_start_call_main (/lib64/libc.so.6+0x3681)
#46 0x00007efe29cf3798 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x3798)
#47 0x0000000000d34be5 _start (/home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin/clang+0xd34be5)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 23.0.0git (https://github.com/kleisauke/llvm-project 8c8189494b548b0fc6ba858701daae043c7b8306)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/kleisauke/emsdk/llvm/git/build_wasm-function-pointers-rebase6_64/bin
Build config: +assertions
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/gsourceclosure-b0b801.c
clang: note: diagnostic msg: /tmp/gsourceclosure-b0b801.sh
clang: note: diagnostic msg: 

********************
```
</details>

turran replied:
Hello @kleisauke 
Thanks for the feedback. Given the current status of the PR, I can build GLib without your fix function pointer patch and only apply a small one. The whole GStreamer works without any function signature mismatch, both static and runtime casts. Let me know if something is still missing. I've also added a documentation file that explains the cases where it is impossible for the compiler to fix, the GLib patch only addresses such situations.

<details>

<summary>The patch</summary>

```diff
diff --git a/glib/glist.c b/glib/glist.c
index 754a3ec52..1e13b4087 100644
--- a/glib/glist.c
+++ b/glib/glist.c
@@ -1097,7 +1097,7 @@ g_list_foreach (GList    *list,
 static GList*
 g_list_insert_sorted_real (GList    *list,
                            gpointer  data,
-                           GFunc     func,
+                           GCompareDataFunc     func,
                            gpointer  user_data)
 {
   GList *tmp_list = list;
@@ -1113,13 +1113,13 @@ g_list_insert_sorted_real (GList    *list,
       return new_list;
     }
   
-  cmp = ((GCompareDataFunc) func) (data, tmp_list->data, user_data);
+  cmp = func (data, tmp_list->data, user_data);
 
   while ((tmp_list->next) && (cmp > 0))
     {
       tmp_list = tmp_list->next;
 
-      cmp = ((GCompareDataFunc) func) (data, tmp_list->data, user_data);
+      cmp = func (data, tmp_list->data, user_data);
     }
 
   new_list = _g_list_alloc0 ();
@@ -1170,7 +1170,7 @@ g_list_insert_sorted (GList        *list,
                       gpointer      data,
                       GCompareFunc  func)
 {
-  return g_list_insert_sorted_real (list, data, (GFunc) func, NULL);
+  return g_list_insert_sorted_real (list, data, (GCompareDataFunc) func, NULL);
 }
 
 /**
@@ -1201,13 +1201,13 @@ g_list_insert_sorted_with_data (GList            *list,
                                 GCompareDataFunc  func,
                                 gpointer          user_data)
 {
-  return g_list_insert_sorted_real (list, data, (GFunc) func, user_data);
+  return g_list_insert_sorted_real (list, data, func, user_data);
 }
 
 static GList *
 g_list_sort_merge (GList     *l1, 
                    GList     *l2,
-                   GFunc     compare_func,
+                   GCompareDataFunc     compare_func,
                    gpointer  user_data)
 {
   GList list, *l, *lprev;
@@ -1218,7 +1218,7 @@ g_list_sort_merge (GList     *l1,
 
   while (l1 && l2)
     {
-      cmp = ((GCompareDataFunc) compare_func) (l1->data, l2->data, user_data);
+      cmp = compare_func (l1->data, l2->data, user_data);
 
       if (cmp <= 0)
         {
@@ -1242,7 +1242,7 @@ g_list_sort_merge (GList     *l1,
 
 static GList * 
 g_list_sort_real (GList    *list,
-                  GFunc     compare_func,
+                  GCompareDataFunc     compare_func,
                   gpointer  user_data)
 {
   GList *l1, *l2;
@@ -1301,7 +1301,7 @@ GList *
 g_list_sort (GList        *list,
              GCompareFunc  compare_func)
 {
-  return g_list_sort_real (list, (GFunc) compare_func, NULL);
+  return g_list_sort_real (list, (GCompareDataFunc) compare_func, NULL);
 }
 
 /**
@@ -1334,7 +1334,7 @@ g_list_sort_with_data (GList            *list,
                        GCompareDataFunc  compare_func,
                        gpointer          user_data)
 {
-  return g_list_sort_real (list, (GFunc) compare_func, user_data);
+  return g_list_sort_real (list, compare_func, user_data);
 }
 
 /**
diff --git a/glib/gslist.c b/glib/gslist.c
index 7d4051c24..120a35363 100644
--- a/glib/gslist.c
+++ b/glib/gslist.c
@@ -890,10 +890,10 @@ g_slist_foreach (GSList   *list,
 }
 
 static GSList*
-g_slist_insert_sorted_real (GSList   *list,
-                            gpointer  data,
-                            GFunc     func,
-                            gpointer  user_data)
+g_slist_insert_sorted_real (GSList               *list,
+                            gpointer             data,
+                            GCompareDataFunc     func,
+                            gpointer             user_data)
 {
   GSList *tmp_list = list;
   GSList *prev_list = NULL;
@@ -910,14 +910,14 @@ g_slist_insert_sorted_real (GSList   *list,
       return new_list;
     }
 
-  cmp = ((GCompareDataFunc) func) (data, tmp_list->data, user_data);
+  cmp = func (data, tmp_list->data, user_data);
 
   while ((tmp_list->next) && (cmp > 0))
     {
       prev_list = tmp_list;
       tmp_list = tmp_list->next;
 
-      cmp = ((GCompareDataFunc) func) (data, tmp_list->data, user_data);
+      cmp = func (data, tmp_list->data, user_data);
     }
 
   new_list = _g_slist_alloc ();
@@ -961,7 +961,7 @@ g_slist_insert_sorted (GSList       *list,
                        gpointer      data,
                        GCompareFunc  func)
 {
-  return g_slist_insert_sorted_real (list, data, (GFunc) func, NULL);
+  return g_slist_insert_sorted_real (list, data, (GCompareDataFunc) func, NULL);
 }
 
 /**
@@ -986,14 +986,14 @@ g_slist_insert_sorted_with_data (GSList           *list,
                                  GCompareDataFunc  func,
                                  gpointer          user_data)
 {
-  return g_slist_insert_sorted_real (list, data, (GFunc) func, user_data);
+  return g_slist_insert_sorted_real (list, data, func, user_data);
 }
 
 static GSList *
-g_slist_sort_merge (GSList   *l1,
-                    GSList   *l2,
-                    GFunc     compare_func,
-                    gpointer  user_data)
+g_slist_sort_merge (GSList               *l1,
+                    GSList               *l2,
+                    GCompareDataFunc     compare_func,
+                    gpointer             user_data)
 {
   GSList list, *l;
   gint cmp;
@@ -1002,7 +1002,7 @@ g_slist_sort_merge (GSList   *l1,
 
   while (l1 && l2)
     {
-      cmp = ((GCompareDataFunc) compare_func) (l1->data, l2->data, user_data);
+      cmp = compare_func (l1->data, l2->data, user_data);
 
       if (cmp <= 0)
         {
@@ -1021,9 +1021,9 @@ g_slist_sort_merge (GSList   *l1,
 }
 
 static GSList *
-g_slist_sort_real (GSList   *list,
-                   GFunc     compare_func,
-                   gpointer  user_data)
+g_slist_sort_real (GSList               *list,
+                   GCompareDataFunc     compare_func,
+                   gpointer             user_data)
 {
   GSList *l1, *l2;
 
@@ -1068,7 +1068,7 @@ GSList *
 g_slist_sort (GSList       *list,
               GCompareFunc  compare_func)
 {
-  return g_slist_sort_real (list, (GFunc) compare_func, NULL);
+  return g_slist_sort_real (list, (GCompareDataFunc) compare_func, NULL);
 }
 
 /**
@@ -1086,7 +1086,7 @@ g_slist_sort_with_data (GSList           *list,
                         GCompareDataFunc  compare_func,
                         gpointer          user_data)
 {
-  return g_slist_sort_real (list, (GFunc) compare_func, user_data);
+  return g_slist_sort_real (list, compare_func, user_data);
 }
 
 /**
```
</details>

kleisauke replied:
@turran That sounds great! Indeed, this may not be necessary after all. For full transparency, I tested commit https://github.com/fluendo/glib/commit/58f4863e42a5a17fa0cc9187e3fd3b330e537909 yesterday, but it didn't seem to work on my end (see e.g. commit https://github.com/kleisauke/wasm-vips/commit/6b788ad97e57100d5002d502cf1be9197656476e). I'll re-test this (and include the stack backtrace if I can still reproduce it).

(Please feel free to resolve this conversation, as I seem unable to do so for some reason).

kleisauke replied:
For context, it was this function signature mismatch:
```console
RuntimeError: function signature mismatch
    at vips.wasm.__wasm_runtime_wrapper_iii_to_iiii____glib_gslist_c_0 (wasm://wasm/vips.wasm-0156d6e6:wasm-function[7739]:0x3366ad)
    at vips.wasm.g_slist_sort_real (wasm://wasm/vips.wasm-0156d6e6:wasm-function[2250]:0xa69f3)
    at vips.wasm.g_slist_sort (wasm://wasm/vips.wasm-0156d6e6:wasm-function[857]:0x30fc0)
    at vips.wasm.vips_object_class_install_argument (wasm://wasm/vips.wasm-0156d6e6:wasm-function[105]:0x9a85)
    at vips.wasm.__vips_object_class_intern_init_vii (wasm://wasm/vips.wasm-0156d6e6:wasm-function[7911]:0x3433e1)
    at vips.wasm.g_type_class_get (wasm://wasm/vips.wasm-0156d6e6:wasm-function[271]:0x1252a)
    at vips.wasm.g_type_class_get (wasm://wasm/vips.wasm-0156d6e6:wasm-function[271]:0x12241)
    at vips.wasm.vips_class_map_all (wasm://wasm/vips.wasm-0156d6e6:wasm-function[658]:0x27714)
    at vips.wasm.vips_class_map_all (wasm://wasm/vips.wasm-0156d6e6:wasm-function[658]:0x277e2)
    at vips.wasm.vips_class_build_hash_cb (wasm://wasm/vips.wasm-0156d6e6:wasm-function[5082]:0x24b0d0)
```

I was able to resolve this in libvips via: https://github.com/libvips/libvips/compare/master...kleisauke:llvm-pr-153168. With that changeset, along with this PR, the wasm-vips test suite now passes. Thank you!

turran replied:
Great news!
(Indeed, the gslist (and glist) needed my small patch)
